File: backend/app/schemas/researcher.py
"""
Pydantic schemas for WeatherLab / Researcher Intelligence.
Covers NWP model comparison (GFS, ECMWF, IMD/WRF), consensus, climate anomalies,
scientific datasets, research reports, and AI research assistant.
"""

from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field


class NWPModelData(BaseModel):
    model_name: str  # GFS, ECMWF, IMD/WRF, ICON
    temperature_c: float
    rainfall_prob_pct: int
    rainfall_mm_24h: float
    wind_speed_kmh: float
    humidity_pct: int
    pressure_hpa: float
    hourly_series: List[Dict[str, Any]] = []


class NWPComparisonResponse(BaseModel):
    location_name: str
    latitude: float
    longitude: float
    forecast_period: str
    models: Dict[str, NWPModelData]
    consensus_level: str  # HIGH, MODERATE, LOW
    consensus_summary: str
    disagreement_analysis: str
    uncertainty_rating: str  # Low, Moderate, High
    uncertainty_explanation: str
    timeline_dates: List[str] = []
    generated_at: str


class ClimateAnomalyMetrics(BaseModel):
    location_name: str
    baseline_period: str  # "1991–2020"
    current_temp_c: float
    current_temp_anomaly_c: float  # e.g. +1.2
    rainfall_season_mm: float
    rainfall_anomaly_pct: float  # e.g. +14.5%
    extreme_events_count_ytd: int
    historical_trend_points: List[Dict[str, Any]] = []
    summary: str


class DatasetItem(BaseModel):
    id: str
    name: str
    coverage: str
    resolution: str
    period: str
    format: str
    source: str
    file_size_mb: Optional[float] = None
    download_url: Optional[str] = None
    parameters: List[str] = []


class ResearchReportRequest(BaseModel):
    title: str
    report_type: str  # Weather Trend Analysis, Rainfall Analysis, Extreme Events, Climate Impact, NWP Intercomparison
    location_name: str
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    period: str
    parameters: List[str] = ["temperature", "rainfall", "wind"]
    include_maps: bool = True
    include_charts: bool = True
    include_nwp: bool = True
    include_findings: bool = True
    custom_notes: Optional[str] = None


class ResearchReportData(BaseModel):
    report_id: str
    title: str
    report_type: str
    location_name: str
    generated_at: str
    executive_summary: str
    study_area: str
    data_sources: List[Dict[str, str]]
    methodology: str
    statistical_summary: Dict[str, Any]
    nwp_intercomparison: Optional[Dict[str, Any]] = None
    climate_trends: Optional[Dict[str, Any]] = None
    key_findings: List[str]
    limitations: List[str]
    citations: List[str]


class WeatherLabSummaryResponse(BaseModel):
    success: bool
    location_name: str
    timestamp: str
    nwp_comparison: NWPComparisonResponse
    climate_anomalies: ClimateAnomalyMetrics
    available_datasets: List[DatasetItem]
    recent_reports: List[Dict[str, Any]]
    active_layers: List[str]

File: frontend/src/types/researcher.ts
/**
 * TypeScript definitions for WeatherLab / Researcher Intelligence.
 */

export interface NWPModelData {
  model_name: string;
  temperature_c: number;
  rainfall_prob_pct: number;
  rainfall_mm_24h: number;
  wind_speed_kmh: number;
  humidity_pct: number;
  pressure_hpa: number;
  hourly_series: {
    date: string;
    temp: number;
    rain: number;
    wind: number;
    humidity: number;
  }[];
}

export interface NWPComparisonResponse {
  location_name: string;
  latitude: number;
  longitude: number;
  forecast_period: string;
  models: {
    GFS: NWPModelData;
    ECMWF: NWPModelData;
    IMD_WRF: NWPModelData;
    [key: string]: NWPModelData;
  };
  consensus_level: 'High' | 'Moderate' | 'Low' | string;
  consensus_summary: string;
  disagreement_analysis: string;
  uncertainty_rating: 'Low' | 'Moderate' | 'High' | string;
  uncertainty_explanation: string;
  timeline_dates: string[];
  generated_at: string;
}

export interface ClimateAnomalyMetrics {
  location_name: string;
  baseline_period: string;
  current_temp_c: number;
  current_temp_anomaly_c: number;
  rainfall_season_mm: number;
  rainfall_anomaly_pct: number;
  extreme_events_count_ytd: number;
  historical_trend_points: {
    year: number;
    anomaly_c: number;
    baseline: number;
  }[];
  summary: string;
}

export interface DatasetItem {
  id: string;
  name: string;
  coverage: string;
  resolution: string;
  period: string;
  format: string;
  source: string;
  file_size_mb?: number;
  download_url?: string;
  parameters: string[];
}

export interface ResearchReportRequest {
  title: string;
  report_type: string;
  location_name: string;
  latitude?: number;
  longitude?: number;
  period: string;
  parameters: string[];
  include_maps: boolean;
  include_charts: boolean;
  include_nwp: boolean;
  include_findings: boolean;
  custom_notes?: string;
}

export interface ResearchReportData {
  report_id: string;
  title: string;
  report_type: string;
  location_name: string;
  generated_at: string;
  executive_summary: string;
  study_area: string;
  data_sources: { name: string; type: string }[];
  methodology: string;
  statistical_summary: Record<string, any>;
  nwp_intercomparison?: Record<string, any>;
  climate_trends?: Record<string, any>;
  key_findings: string[];
  limitations: string[];
  citations: string[];
}

export interface WeatherLabSummary {
  success: boolean;
  location_name: string;
  timestamp: string;
  nwp_comparison: NWPComparisonResponse;
  climate_anomalies: ClimateAnomalyMetrics;
  available_datasets: DatasetItem[];
  recent_reports: {
    id: string;
    title: string;
    date: string;
    format: string;
    size: string;
  }[];
  active_layers: string[];
}
